src/main/java/ch/njol/skript/expressions/ExprBeaconValues.java

```diff
@@ -0,0 +1,237 @@
+package ch.njol.skript.expressions;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.classes.Changer.ChangeMode;
+import ch.njol.skript.doc.*;
+import ch.njol.skript.expressions.base.EventValueExpression;
+import ch.njol.skript.expressions.base.PropertyExpression;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.util.Getter;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+import com.destroystokyo.paper.event.block.BeaconEffectEvent;
+import io.papermc.paper.event.block.BeaconActivatedEvent;
+import io.papermc.paper.event.block.BeaconDeactivatedEvent;
+import io.papermc.paper.event.player.PlayerChangeBeaconEffectEvent;
+import org.bukkit.block.Beacon;
+import org.bukkit.block.Block;
+import org.bukkit.event.Event;
+import org.bukkit.potion.PotionEffectType;
+import org.jetbrains.annotations.Nullable;
+
+import static ch.njol.util.Math2.floor;
+
+@Name("Beacon Effects")
+@Description({
+	"The active effects of the beacon.",
+	"The secondary effect can only be set to regeneration and it can only be set when the beacon has achieved the highest tier.",
+	"You can only change the range on Paper."
+})
+@Examples({
+	"broadcast tier of {_block}",
+	"set primary beacon effect of {_block} to haste",
+	"add 1 to range of {_block}"
+})
+@Since("INSERT VERSION")
+public class ExprBeaconValues extends PropertyExpression<Block, Object> {
+
+	enum BeaconValues {
+		PRIMARY("primary beacon effect"),
+		SECONDARY("secondary beacon effect"),
+		RANGE("[beacon] range"),
+		TIER("[beacon] tier");
+
+		private final String name;
+
+		BeaconValues(String name) {
+			this.name = name;
+		}
+	}
+
+	private static boolean PAPER_EVENTS, PAPER_RANGE = false;
+	private static BeaconValues[] beaconValues = BeaconValues.values();
+
+	static {
+		if (Skript.classExists("com.destroystokyo.paper.event.block.BeaconEffectEvent"))
+			PAPER_EVENTS = true;
+		if (Skript.methodExists(Beacon.class, "getEffectRange"))
+			PAPER_RANGE = true;
+
+		int size = beaconValues.length;
+		String[] patterns = new String[size * 2];
+		for (BeaconValues value : beaconValues) {
+			patterns[2 * value.ordinal()] = "%blocks%['s] " + value.name;
+			patterns[2 * value.ordinal() + 1] = value.name + " [of %blocks%]";
+		}
+
+		Skript.registerExpression(ExprBeaconValues.class, Object.class, ExpressionType.PROPERTY, patterns);
+	}
+
+	private BeaconValues valueType;
+	private boolean isEffect;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		valueType = beaconValues[(int) floor(matchedPattern/2)];
+		if (valueType == BeaconValues.RANGE && !PAPER_RANGE) {
+			Skript.error("This can only be used on Paper.");
+			return false;
+		}
+		if (exprs[0] != null) {
+			setExpr((Expression<Block>) exprs[0]);
+		} else {
+			if (!PAPER_EVENTS)
```

Comment:
It might be better to add the [] to the patterns only if paper_events exist, rather than checking here.